sdks/java/io/hadoop-input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java

```diff
@@ -137,27 +153,74 @@ public void readUsingHadoopInputFormat() {
                 .withStatement(String.format("insert into %s values(?, ?)", tableName))
                 .withPreparedStatementSetter(new PrepareStatementFromTestRow()));
 
-    writePipeline.run().waitUntilFinish();
+    PipelineResult writeResult = writePipeline.run();
+    writeResult.waitUntilFinish();
 
     PCollection<String> consolidatedHashcode =
         readPipeline
             .apply(
                 "Read using HadoopInputFormat",
                 HadoopInputFormatIO.<LongWritable, TestRowDBWritable>read()
                     .withConfiguration(hadoopConfiguration.get()))
+            .apply("Collect read time", ParDo.of(new TimeMonitor<>(NAMESPACE, "read_time")))
             .apply("Get values only", Values.create())
             .apply("Values as string", ParDo.of(new SelectNameFn()))
             .apply("Calculate hashcode", Combine.globally(new HashingFn()));
 
     PAssert.thatSingleton(consolidatedHashcode).isEqualTo(getExpectedHashForRowCount(numberOfRows));
 
-    readPipeline.run().waitUntilFinish();
+    PipelineResult readResult = readPipeline.run();
+    readResult.waitUntilFinish();
+
+    collectAndPublishMetrics(writeResult, readResult);
+  }
+
+  private void collectAndPublishMetrics(PipelineResult writeResult, PipelineResult readResult) {
+    String uuid = UUID.randomUUID().toString();
+    String timestamp = Timestamp.now().toString();
+
+    Set<Function<MetricsReader, NamedTestResult>> readSuppliers = getReadSuppliers(uuid, timestamp);
+    Set<Function<MetricsReader, NamedTestResult>> writeSuppliers =
+        getWriteSuppliers(uuid, timestamp);
+
+    IOITMetrics readMetrics =
+        new IOITMetrics(readSuppliers, readResult, NAMESPACE, uuid, timestamp);
+    IOITMetrics writeMetrics =
+        new IOITMetrics(writeSuppliers, writeResult, NAMESPACE, uuid, timestamp);
+    readMetrics.publish(bigQueryDataset, bigQueryTable);
+    writeMetrics.publish(bigQueryDataset, bigQueryTable);
+  }
+
+  private Set<Function<MetricsReader, NamedTestResult>> getWriteSuppliers(
+      String uuid, String timestamp) {
+    Set<Function<MetricsReader, NamedTestResult>> suppliers = new HashSet<>();
+    suppliers.add(
+        reader -> {
+          long writeStart = reader.getStartTimeMetric("write_time");
+          long writeEnd = reader.getEndTimeMetric("write_time");
+          return NamedTestResult.create(
+              uuid, timestamp, "write_time", (writeEnd - writeStart) / 1e3);
+        });
+    return suppliers;
+  }
+
+  private Set<Function<MetricsReader, NamedTestResult>> getReadSuppliers(
+      String uuid, String timestamp) {
+    Set<Function<MetricsReader, NamedTestResult>> suppliers = new HashSet<>();
+    suppliers.add(
+        reader -> {
+          long readStart = reader.getStartTimeMetric("read_time");
+          long readEnd = reader.getEndTimeMetric("read_time");
+          return NamedTestResult.create(uuid, timestamp, "write_time", (readEnd - readStart) / 1e3);
```

Comment:
That was a tricky one! :D I believe there should be "read_time" otherwise we will save "write_time" twice in the bq

Reply:
Nice catch, fixed :) MongoDBIOIT failed because of some Mongo problems, unrelated to metrics.